### PR TITLE
Best practice solution sample using X.509 authentication

### PR DESCRIPTION
**Pull request by schoims**

### CODE REVIEW

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Samples
+{
+    /// <summary>
+    /// Demonstrates how to register a device with the device provisioning service using a X.509 leaf certificate, and then
+    /// use the registration information to authenticate to IoT Hub.
+    /// </summary>
+    internal class BestPracticeSampleX509
```
Comment (schoims) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:24]: This file combined X509Sample and DeviceReconnectionSample.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 2]
```diff
@@ -0,0 +1,533 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Client.Exceptions;
+using Microsoft.Azure.Devices.Client.Samples;
+using Microsoft.Azure.Devices.Provisioning.Client.Transport;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Extensions.Logging;
+using System;
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:10]: sort/remove

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 3]
```diff
@@ -0,0 +1,533 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:1]: This sample looks like a copy of the DeviceReconnectionSample, but I wonder if all the extra fluff is required to demonstrate how to authenticate with x509 to DPS and then hub? It inflates the size of this sample quite a bit. Is that desirable?

Seems to me we'd simply need to authenticate and then send some telemetry, and then include a link to the device reconnection sample to say that for a device app to be proper, it needs to include other concerns like client lifetime, connection status monitoring, etc.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.csproj [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+    <LangVersion>9.0</LangVersion>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <LangVersion>9.0</LangVersion>
+  </PropertyGroup>
+
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.csproj:17]: Not sure what this is.

```suggestion
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1 [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+$resourceGroupExists = az group exists -n $resourceGroup
+if ($resourceGroupExists -ne $true){
+    Write-Host "Resource Group '$resourceGroup' does not exist. Exiting..."
+    exit
+}
+
+# Check if the DPS instance exists. If not, exit.
+$dpsExists = az iot dps show --name $dpsName -g $resourceGroup 2>nul
+if ($dpsExists -eq $null){
+    Write-Host "DPS '$dpsName' does not exist under '$resourceGroup'. Exiting..."
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1:35]: Why exit? Could be the other resources are still there.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        private const int MaxRetryCount = int.MaxValue;
+        private const int MaxExponent = 22; // Avoid integer overlow (max of 30) and clamp max wait to just over 1 hour (2^22 = 1.16 hours).
+
+        private readonly Random _rng = new();
+        private readonly object _randLock = new();
+        private readonly HashSet<Type> _exceptionsToBeRetried;
+
+        private readonly ILogger _logger;
+
+        internal CustomRetryPolicy(HashSet<Type> exceptionsToBeRetried, ILogger logger)
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs:26]: Why does this sample require a custom retry policy?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/GenerateGroupEnrollment.ps1 [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    # Deletes and recreate enrollmentGroup if it already exists
+    [Parameter(Mandatory=$false)]
+    [switch] $force = $false
+)
+# GroupEnrollment Id created to use in sample.
+$groupEnrollmentId = "x509GroupEnrollment"
+
+# Check certificate file extension.
+if (($intermediateCertName.EndsWith('.pem') -ne $true) -and ($intermediateCertName.EndsWith('.cer') -ne $true)){
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/GenerateGroupEnrollment.ps1:25]: Missing a space in all your if statements.

```suggestion
if (($intermediateCertName.EndsWith('.pem') -ne $true) -and ($intermediateCertName.EndsWith('.cer') -ne $true)) {
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/GenerateGroupEnrollment.ps1 [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+if ($dpsExists -eq $null){
+    Write-Host "Dps '$dpsName' does not exist under '$resourceGroup'. Exiting..."
+    exit
+}
+
+# Check if the enrollment group already exists in dps instance. If it does, delete and regenerate the group enrollment.
+Write-Host "`Checking if '$groupEnrollmentId' enrollment group already exists in '$dpsName'..."
+$groupEnrollmentExists = az iot dps enrollment-group show --dps-name $dpsName -g $resourceGroup --enrollment-id $groupEnrollmentId 2>nul
+if ($groupEnrollmentExists)
+{
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/GenerateGroupEnrollment.ps1:48]: I prefer braces on a new line like this one, but for consistency you should choose one style and stick with it.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1 (hunk 5 above)
Comment (schoims) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1:35]: This is to check if the dps instance exists before deleting the enrollment group created earlier.  If there is no dps instance, then no enrollment group to delete.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/GenerateGroupEnrollment.ps1 [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    if ($force) {
+        Write-Host "Deleting existing enrollment group '$groupEnrollmentId' in '$dpsName'..."
+        az iot dps enrollment-group delete -g $resourceGroup --eid $groupEnrollmentId --dps-name $dpsName
+        Write-Host "Enrollment group '$groupEnrollmentId' is deleted in '$dpsName'."
+    }
+    else {
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/GenerateGroupEnrollment.ps1:54]: Same here.

Either
```csharp
}
else
{
```

or

```csharp
} else {
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/Program.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+            try
+            {
+                var sample = new BestPracticeSampleX509(parameters.GlobalDeviceEndpoint, parameters.IdScope, parameters.CertificateName, parameters.CertificatePassword, parameters.TransportType, logger, parameters);
+                await sample.RunSampleAsync(runningTime);
+            }
+            catch (OperationCanceledException)
+            { logger.LogDebug("OperationCanceledException caught."); }
+            catch (Exception ex)
+            { logger.LogDebug($"Exception caught. \n{ex}"); }
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/Program.cs:57]: Observe standard bracing.

```suggestion
            catch (OperationCanceledException)
            {
                logger.LogDebug("OperationCanceledException caught.");
            }
            catch (Exception ex)
            {
                logger.LogDebug($"Exception caught. \n{ex}");
            }
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/X509AuthSetup.ps1 [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    [Parameter(Mandatory=$true)]
+    [string] $dpsName,
+
+    # Device Id created to use in sample.
+    [Parameter(Mandatory=$true)]
+    [string] $deviceId
+)
+
+# Check that script is run in admin mode
+$isAdmin = ([Security.Principal.WindowsPrincipal] [Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole] "Administrator")
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/X509AuthSetup.ps1:27]: Why is this required? Please add a code comment.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs (hunk 3 above)
Comment (schoims) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:1]: The purpose of combining two samples is to give a full solution. I thought it would be nicer if users can see the full solution instead of having to open another sample and combine them themselves. But I see your point.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
+# Best Practice Solution Sample with X.509 Authentication
+
+## Objective
+
+This doc demonstrates a real-world example to help you get started with building your own custom IoT cloud solution with X.509 certificate-based authentication. See [here](https://learn.microsoft.com/en-us/azure/iot-dps/concepts-service#Authentication-mechanism) to review available attestation mechanisms.
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:5]: Do not include the language in a URL.

```suggestion
This doc demonstrates a real-world example to help you get started with building your own custom IoT cloud solution with X.509 certificate-based authentication. See [here](https://learn.microsoft.com/azure/iot-dps/concepts-service#Authentication-mechanism) to review available attestation mechanisms.
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+## Example
+
+Company-X is a manufacturing company that wants to manage various types of products remotely.
+
+The business needs of Company-X are -
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:11]:
```suggestion
The business needs of Company-X are:
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+- provision devices securely at scale
+- group devices by production line
+- collect data from devices
+- manage and monitor devices remotely (e.g., controlling the temperature of a thermostat)
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:16]:
```suggestion
- Provision devices securely at scale.
- Group devices by production line.
- Collect data from devices.
- Manage and monitor devices remotely (e.g., controlling the temperature of a thermostat).
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+### Prerequisites
+
+1. Create a resource group and one or more IoT hubs.
+2. Set up an IoT hub Device Provisioning Service (DPS) instance.
+3. Link the IoT hub(s) to the DPS instance.
+See [here](https://learn.microsoft.com/en-us/azure/iot-dps/quick-setup-auto-provision) for the instructions.
+
+In this solution, we will use the standard X.509 CA certificate authentication.
+
+### 1. Determine the Public Key Infrastructure (PKI) design
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:33]: Don't put numbered bullets in headers. Either use bullets or headers.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+| Purchase a CA certificate from a commercial CA and issue certificates within the organization from internal, self-managed CAs that are chained to the external root CA | For production environments if your devices will interact with third-party products or services |
+
+In this solution, we will use a self-managed PKI. If you want to see a sample that registers a device to your DPS instance using a self-signed device certificate, see [here](https://github.com/Azure/azure-iot-sdk-csharp/tree/main/provisioning/device/samples/Getting%20Started/X509Sample).
+
+### 2. Generate self-signed X.509 certificates and verify root certificate
+
+1. Create a secure string password to use in the creation of self-signed certificates. Open Windows PowerShell as administrator.
+
+```powershell
+    $password= ConvertTo-SecureString <your password> -AsPlainText -Force
```
Comment (drwill-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:49]:
```suggestion
    $password = ConvertTo-SecureString <your password> -AsPlainText -Force
```

**Review by drwill-ms — CHANGES_REQUESTED** [review]
Comments inline.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs (hunk 6 above)
Comment (schoims) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs:26]: This was copied from DeviceReconnectionSample. Do you suggest removing this?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+![solution](media/auth_flow_diagram.png)
+> **Note**\
+> The Device Process is the device (host) application that holds the provisioning client and IoT hub communication code.
+
+### Prerequisites
+
+1. Create a resource group and one or more IoT hubs.
+2. Set up an IoT hub Device Provisioning Service (DPS) instance.
+3. Link the IoT hub(s) to the DPS instance.
+See [here](https://learn.microsoft.com/en-us/azure/iot-dps/quick-setup-auto-provision) for the instructions.
```
Comment (patilsnr) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:29]: Another place to delete `en-us` from the link

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+In this solution, we will use the standard X.509 CA certificate authentication.
+
+### 1. Determine the Public Key Infrastructure (PKI) design
+
+See [here](https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2012-R2-and-2012/dn786436(v=ws.11)) for Public Key Infrastructure (PKI) design options.
```
Comment (patilsnr) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:35]: `en-us` again

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        -rootCertPassword $password `
+        -dpsResourceGroup <DPS instance resource group> `
+        -dpsName <DPS instance name> `
+        -deviceId <device Id>
+```
+
+[X509AuthSetup.ps1](https://github.com/Azure/azure-iot-sdk-csharp/tree/main/provisioning/device/samples/solutions/BestPracticeSampleX509/X509AuthSetup.ps1)  first issues a self-signed root CA certificate. Then it uses the root CA certificate to generate a unique intermediate certificate for each product line. Finally, it uses the production line certificate, to generate a unique device (end-entity) certificate for each device manufactured on the line.
+
+> **Note**\
+> Read more about X.509 certificate attestation [here](https://learn.microsoft.com/en-us/azure/iot-dps/concepts-x509-attestation).
```
Comment (patilsnr) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:66]: `en-us`

Comment (patilsnr) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:66]: same below

**Review by patilsnr — COMMENTED** [review]
I like where this is going! I think the team could benefit from a meeting where we go over the functionality of this sample. Great job so far!

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md (hunk 12 above)
Comment (abhipsaMisra) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:5]:
```suggestion
This doc demonstrates a real-world example to help you get started with building your own custom IoT cloud solution with X.509 certificate-based authentication. See [here](https://learn.microsoft.com/azure/iot-dps/concepts-service#attestation-mechanism) to review available attestation mechanisms.
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 20]
```diff
@@ -0,0 +1,128 @@
+# Best Practice Solution Sample with X.509 Authentication
+
+## Objective
+
+This doc demonstrates a real-world example to help you get started with building your own custom IoT cloud solution with X.509 certificate-based authentication. See [here](https://learn.microsoft.com/en-us/azure/iot-dps/concepts-service#Authentication-mechanism) to review available attestation mechanisms.
+
+## Example
+
+Company-X is a manufacturing company that wants to manage various types of products remotely.
```
Comment (timtay-microsoft) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:9]: "Contoso" is the usual go-to fictional company Microsoft uses in cases like this, FYI

File: provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1 (hunk 5 above)
Comment (schoims) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1:35]: Instead of exiting here, I can skip deleting the enrollment group if the dps instance doesn't exist and then just delete the device (if it exists).

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md (hunk 20 above)
Comment (schoims) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:9]: Sure!

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs (hunk 1 above)
Comment (drwill-ms) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:24]: Meaning we got rid of the device connection sample, or just used it as a base?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs (hunk 3 above)
Comment (drwill-ms) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:1]: Perhaps we should have a team talk about it, but I feel this is too much. We don't do the same thing for every sample because it dilutes what a user should be learning by reading a sample. We want most of them focused on just 1 thing.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1 (hunk 5 above)
Comment (drwill-ms) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1]: I mean, maybe change the logic from not exists to exists, and delete inside the braces if the things you need are there?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs (hunk 6 above)
Comment (drwill-ms) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs:26]: Per the earlier comment about keeping this sample focused, yes, I would remove most of the superfluous stuff.

Comment (andyk-ms) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs:26]: Wait, can we discuss about this? I think as a best practice, we should include retry and exception handling. The reason is that it is not difficult to implement the happy path. Most of the customer support cases are due to weak handling of exceptions.

Take a look at the customRetry from V1 DeviceReconnectionSample I have patched up for testing with ABB:
https://github.com/Azure/azure-iot-sdk-csharp/tree/andyk/abb

File: provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1 (hunk 5 above)
Comment (schoims) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/Cleanup.ps1]: Makes sense! Thank you for the suggestion and the comments!

File: provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+                lock (_randLock)
+                {
+                    int plusOrMinus = _rng.Next(0, 2) * 2 - 1;
+
+                    // a random double from 0 to 999, positive or negative
+                    jitterMs =  plusOrMinus * _rng.NextDouble() * 1000;
+                }
+
+                // Avoid integer overlow and clamp max wait.
+                int exponent = Math.Min(MaxExponent, currentRetryCount);
```
Comment (andyk-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs:59]: exponents less than 10 essentially result in same retry interval.
```suggestion
                int exponent = Math.Min(MaxExponent, currentRetryCount+MinExponent);
```
Add this MinExponent at the beginning (value from 7-10 works better):
`private const int MinExponent = 8;`

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs (hunk 1 above)
Comment (andyk-ms) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:24]: Maybe we should have the design reviewed to determine what feature we should include.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 22]
```diff
@@ -0,0 +1,533 @@
...
+                var provClient = ProvisioningDeviceClient.Create(
+                    _globalEndpoint,
+                    _idScope,
+                    security,
+                    transport);
+
+                Console.WriteLine($"Initialized for registration Id {security.GetRegistrationID()}.");
+
+                Console.WriteLine("Registering with the device provisioning service... ");
+                DeviceRegistrationResult result = await provClient.RegisterAsync(s_appCancellation.Token);
```
Comment (andyk-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:113]: If we will include retry logic, we may want to have the same for provisioning.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 23]
```diff
@@ -0,0 +1,533 @@
...
+
+                s_initSemaphore.Dispose();
+                if (_auth != null)
+                {
+                    _auth.Dispose();
+                }
+                s_appCancellation.Dispose();
+            }
+
+            Console.WriteLine("Finished.");
```
Comment (andyk-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:159]: should we use logger?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+2. To build the sample application using dotnet, from terminal navigate to the sample folder (where the .csproj file lives). Then execute the following command and check for build errors:
+
+```powershell
+    dotnet build
+```
+
+3. Set the IdScope, device certificate, and certificate password.
+
+```powershell
+    dotnet run --s <IdScope> --c <CertificateName> --p <your password>
```
Comment (andyk-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:100]: what about transport type?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 25]
```diff
@@ -0,0 +1,533 @@
...
+        private readonly Parameters _parameters;
+
+        private readonly string _globalEndpoint;
+        private readonly string _idScope;
+        private readonly string _certificateName;
+        private string _certificatePassword;
+        private DeviceAuthenticationWithX509Certificate _auth;
+        private string _assignedIoTHub;
+
+        public BestPracticeSampleX509(string globalEndpoint, string idScope, string certificateName, string certificatePassword, TransportType transportType, ILogger logger, Parameters parameters)
```
Comment (andyk-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:68]: maybe debatable, if "parameters" are send, maybe there is no need to send the individual parameter (globalEndpoint, idScope, certificateName ....)
```suggestion
        public BestPracticeSampleX509(ILogger logger, Parameters parameters)
```

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 26]
```diff
@@ -0,0 +1,533 @@
...
+            Console.CancelKeyPress += (sender, eventArgs) =>
+            {
+                eventArgs.Cancel = true;
+                s_appCancellation.Cancel();
+                _logger.LogWarning("Sample execution cancellation requested; will exit.");
+            };
+
+            _logger.LogInformation($"Sample execution started, press Control+C to quit the sample.");
+
+            Console.WriteLine("Loading the certificate...");
```
Comment (andyk-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:95]: use logger?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 27]
```diff
@@ -0,0 +1,533 @@
...
+            try
+            {
+                using X509Certificate2 certificate = LoadProvisioningCertificate();
+
+                using var security = new SecurityProviderX509Certificate(certificate);
+                Console.WriteLine("Initializing the device provisioning client...");
```
Comment (andyk-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:101]: use logger?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs [hunk 28]
```diff
@@ -0,0 +1,533 @@
...
+            _globalEndpoint = globalEndpoint;
+            _idScope = idScope;
+            _certificateName = certificateName;
+            _certificatePassword = certificatePassword;
+            _parameters = parameters;
+        }
+
+        private static bool IsDeviceConnected => s_connectionStatus == ConnectionStatus.Connected;
+
+        public async Task RunSampleAsync(TimeSpan applicationRunningTime)
```
Comment (andyk-ms) [inline provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs:83]: We should discuss about the design - right now, it is a long piece of code process that does not coorespond to the readme or the steps in the diagram. Based on the diagram in readme, it illustrated at least these 3 steps:
1) "getting the certificate", we may show a functional module to 'read from file system', but customer can replace it with 'read a secure stream from openssl client' for example (or other can use a TPM based approach).
2) 'register with DPS with certificate' - another function or can we call another sample that does exactly that? (modular samples building blocks)
3) 'device messaging with hub' - another function focus on processing messages or can we call another sample?

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md (hunk 24 above)
Comment (schoims) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:100]: They are optional parameters and set to default values.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs (hunk 6 above)
Comment (drwill-ms) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs:26]: > Wait, can we discuss about this?

Of course. That's what we're doing and what I called for in another comment. I'm not trying to dictate here. :)


If every sample includes all best practices, customers can't easily figure out what a specific sample is demonstrating. There is just too much code.

If we did that, we might as well just have 1 uber sample that does everything. What's the point?

The only thing I'd consider adding for this sample (in contrast to what the DeviceReconnectionSample has) is that it is recommended for devices to cache their registration information and try to use it first before calling register again (to avoid dps throttling). That would mean implement part of the connection status callback to know when the client should be reinitialized, and the client re/initialize logic, plus the load-from-cache and save-to-cache logic.

File: provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md (hunk 24 above)
Comment (drwill-ms) [reply provisioning/device/samples/solutions/BestPracticeSampleX509/readme.md:100]: Our sample allows for global endpoint override, so this script should too.

**Comment by schoims** [comment]
Closing as we decided to introduce best practice solution samples in v2 first.
